azure-quantum/azure/quantum/target/target_factory.py

```diff
@@ -0,0 +1,94 @@
+##
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+##
+import warnings
+from typing import Dict, List, TYPE_CHECKING, Union
+from azure.quantum.target import *
+
+if TYPE_CHECKING:
+    from azure.quantum import Workspace
+    from azure.quantum._client.models import TargetStatus
+
+# Target ID keyword for parameter-free solvers
+PARAMETER_FREE = "parameterfree"
+
+
+def get_all_target_cls() -> Dict[str, Target]:
+    """Get all target classes by target name"""
+    return {
+        name: _t for t in Target.__subclasses__()
+        for _t in [t] + t.__subclasses__()
+        for name in _t.target_names
+    }
+
+
+class TargetFactory:
+    """Factory class for generating a Target based on a provider and target name"""
+    def __init__(self, workspace: "Workspace"):
+        self.workspace = workspace
```

Comment:
`Factory` classes tend to be singletons, however, here you create one Factory per Workspace. 
Maybe instead of creating a factory per workspace we push the `workspace` s a parameter of `get_targets`?